### ISSUE TITLE
feature request: Vector for log management

### ISSUE THREAD

**Issue by polarathene** [docker-mailserver/docker-mailserver#3561]
### Context

[`Vector`](https://vector.dev/) for easier and better documented configuration, offering more flexible log management than the current `rsyslog` package supports (_syslog focused only_):
- Forwarding logs to external instance of Fail2Ban: https://github.com/docker-mailserver/docker-mailserver/issues/3178#issuecomment-1477101386
- Similar brief mention of adopting Vector for better log management with Redis and Rspamd (_also briefly touches on syslog support, and some other redis config to consider unrelated to logging_):
  - https://github.com/docker-mailserver/docker-mailserver/issues/3138#issuecomment-1453519599
  - https://github.com/docker-mailserver/docker-mailserver/issues/3138#issuecomment-1454251202
  - https://github.com/docker-mailserver/docker-mailserver/pull/3143#pullrequestreview-1324908603
- Process to approach the change discussed here: https://github.com/docker-mailserver/docker-mailserver/pull/3131#issuecomment-1452768189
- Overview / Comparision to `rsyslog` for why to adopt Vector detailed here: https://github.com/docker-mailserver/docker-mailserver/pull/3134#issuecomment-1450935046

Another maintainer approving interest in adopting Vector: https://github.com/docker-mailserver/docker-mailserver/pull/3134#issuecomment-1451031308

### Description

This has been considered for a while. I just haven't properly documented / discussed an issue for going forward.

It's not likely to be a quick and easy change to implement, tests will likely need to be updated and current logging configuration reviewd/revised. Docs should also be added to better communicate the support to users that want to leverage it.

### Alternatives

I think there was some improvements that could be done with `rsyslog` config, but beyond that there were not many other appealing options within a container that I came across. 

Vector looks very nice but needs further investigation to verify as a suitable fit for the project that actually provides real value to ship with the image without increasing burden to maintainers.

### Applicable Users

Anyone that needs better control of log management than we currently have. I am particularly interested in it as our current docs are lacking in this area, our log support with DMS is rather limited AFAIK.

IIRC, there was a concern with how `supervisord` handled logging which @georglauterbach raised upstream but his fix was rejected. I believe this was [related to ANSI-based coloured logs](https://github.com/docker-mailserver/docker-mailserver/pull/1758) and a workaround was implemented at least for `docker logs` stdout. With Vector the log level can be treated as metadata or streams, similar to what syslog supports, but IIRC syslog was less flexibility on "facilities"/tagging for grouping streams, especially when handling multi-line logs.

### Are you going to implement it?

Yes, because I know the probability of someone else doing it is low and I can learn from it.

### What are you going to contribute?

If this is ever contributed, it's probably going to be handled by me 😅 

This is a low priority task due to effort required and ROI for users + maintainers vs other tasks in my backlog. I'm raising this FR for the benefit of others that may want to try tackle it if I don't get around to it.

For reference [this PR](https://github.com/docker-mailserver/docker-mailserver/pull/116) adjusted services to centralize logging to `/var/log/mail` directory (_which we support as a [volume to persist for logs](https://github.com/docker-mailserver/docker-mailserver/blob/bd96c1161e9f9b32bb0b3aaec4b930980a56945f/compose.yaml#L20)_).
- Later `supervisord` was integrated and introduced another location that services output log files to (_although these often appear empty / minimal as services get configured to log elsewhere instead of stdout?_).
- The fail2ban log for services I believe also differed since, with [current monitored services enabled here](https://github.com/docker-mailserver/docker-mailserver/blob/bd96c1161e9f9b32bb0b3aaec4b930980a56945f/target/fail2ban/jail.local).

---

There are some concerns (_as noted by the linked context resources_).
- [Syslog Sink support](https://github.com/vectordotdev/vector/issues/6863#issuecomment-1590253148) should probably be implemented first (many existing services in DMS can easily forward logs into Vector this way, and `rsyslog` will probably be replaced as a result_). Until then you can forward from `rsyslog`, but I recall some issues with that [like this issue](https://github.com/vectordotdev/vector/issues/18643).
- I recall the binary size for Vector being a bit big (_it bundles many features that can be optionally compiled and we probably don't need many of them_), depends how much the added weight to image size matters for approval... May require a custom build if that's the case 🤷‍♂️
  - Some [basic tunables share some size comparisons here](https://github.com/vectordotdev/vector/issues/309#issuecomment-586505606)
  - Feature opt-out would need to be experimented with the `Cargo.toml`? (_seems there is some [`cargo build` flag support](https://github.com/vectordotdev/vector/discussions/9010#discussioncomment-1273630), and building with vendored (external?) deps_)
  - Size was [less than 15MB back in 2019](https://github.com/vectordotdev/vector/pull/689), and could be compressed (_such as [with UPX](https://github.com/vectordotdev/vector/issues/1797)_) for additional reduction.

**Comment by polarathene** [comment]
Relevant changes to consider (_dug up from a local branch experiment back in March 2023_). I probably have some notes elsewhere laying around on my disk, if I find anything else that's relevant I'll update it here 👍 

## Including Vector

While there is no official package in Debian, [Vector supports a variety of install methods](https://vector.dev/docs/setup/installation/operating-systems/debian/) including their own third-party repo if preferrable.

Since we have `packages.sh` that may be a good approach. Alternatively, especially for ease of use (_by not invalidating the `packages.sh` layer during implementation_), one could copy the binary over like we do with the ClamAV database.

**`Dockerfile`:**

```Dockerfile
# Copy binary over from published DockerHub release:
# - https://hub.docker.com/r/timberio/vector
# - NOTE: Adjust the tag to current release tag (eg `:0.33.0-debian`), otherwise it won't be properly version pinned:
# - Besides `-debian` variant, they also offer `-alpine`, `-distroless-static`, `-distroless-libc`
# - They also now have `0.33.X-variant` and `0.X-variant` which will provide the latest minor or patch release
COPY --link --from=docker.io/timberio/vector:0.X-debian /usr/bin/vector /usr/bin/vector
```

The image can be deployed as a separate container for side-car usage with the `/var/log/mail` directory and file `source` config, but since metadata is lost (_especially for `/var/log/mail/mail.log`_), this isn't as helpful. Vector can configure some parsing to infer log level of some entries based on content, same with service and facility. [Vector docs for Docker image deployment](https://vector.dev/docs/setup/installation/platforms/docker/).

---

## Stashed changes

### Rsyslog

This was probably just experimenting with `rsyslog` config. It was a stashed change with the only context "rsyslog file source".

**`target/scripts/build/dms.conf`:**

```
module(load="imfile" Mode="inotify") #needs to be done just once

# File 1
input(type="imfile"
      File="/var/log/supervisor/postfix.log"
      Tag="tag1"
      Facility="mail")

input(type="imfile"
      File="/var/log/supervisor/postgrey.log"
      Facility="mail")
```

### Amavis

There was also a similar one where it looks like I was experimenting with Amavis log configuration (_applied to [this DMS config](https://github.com/docker-mailserver/docker-mailserver/blob/bd96c1161e9f9b32bb0b3aaec4b930980a56945f/target/amavis/conf.d/49-docker-mailserver#L9-L10)_):

```
$log_level = 4;
# $LOGDIR = "/var/log";
# $logfile = "/dev/stdout";
# $do_syslog = 0;
```

---

## Test suite adjustment

A change for tests to be more explicit about services being searched via their own log instead of generic `/var/log/mail/mail.log` output, such as a supervisord stdout log file.

Probably a bit of changes since, but gives a rough overview of approach taken to better focus the scope of log content (_helpful for troubleshooting_).

### Setup

Add these log vars to [this `setup.bash` helper method](https://github.com/docker-mailserver/docker-mailserver/blob/bd96c1161e9f9b32bb0b3aaec4b930980a56945f/test/helper/setup.bash#L18) (_`test/test_helper/common.bash` is the legacy path as well until all tests are converted over to `test/helper/setup.bash`_):

```bash
# Log files for tests to grep:
export LOGFILE_AMAVIS=/var/log/mail/mail.log
export LOGFILE_POSTFIX=/var/log/supervisor/postfix.log
export LOGFILE_POSTGREY=/var/log/supervisor/postgrey.log
```

### Helpers

This is specifically interested in Postfix logs: `/var/log/mail.log` => `"${LOGFILE_POSTFIX}"`

https://github.com/docker-mailserver/docker-mailserver/blob/bd96c1161e9f9b32bb0b3aaec4b930980a56945f/test/helper/sending.bash#L74

### Amavis

Replace all 4 `/var/log/mail/mail.log` instances with `${LOGFILE_AMAVIS}`:

https://github.com/docker-mailserver/docker-mailserver/blob/bd96c1161e9f9b32bb0b3aaec4b930980a56945f/test/tests/parallel/set3/mta/smtp_delivery.bats#L227-L248

### Postfix

This would require the following addition to `main.cf` (_detailed in orignal FR description above regarding [info from this PR](https://github.com/docker-mailserver/docker-mailserver/pull/3134#discussion_r1123919663)_):

```
# Logging (instead of syslog default, output to stdout for supervisor)
maillog_file = /dev/stdout
```

Only requires Postfix logs: `/var/log/mail/mail.log` => `"${LOGFILE_POSTFIX}"`

`virtual_transport` alternative support: 

https://github.com/docker-mailserver/docker-mailserver/blob/bd96c1161e9f9b32bb0b3aaec4b930980a56945f/test/tests/parallel/set3/mta/lmtp_ip.bats#L46

All 6 instances here (_1st one requires quotes to `grep` to adjust to double quotes for variable expansion_):

https://github.com/docker-mailserver/docker-mailserver/blob/bd96c1161e9f9b32bb0b3aaec4b930980a56945f/test/tests/parallel/set3/mta/smtp_delivery.bats#L147-L198

And this one:

https://github.com/docker-mailserver/docker-mailserver/blob/bd96c1161e9f9b32bb0b3aaec4b930980a56945f/test/tests/parallel/set3/mta/smtp_delivery.bats#L221-L225

`PERMIT_DOCKER` test (_also need to adjust quotes_):

https://github.com/docker-mailserver/docker-mailserver/blob/bd96c1161e9f9b32bb0b3aaec4b930980a56945f/test/tests/serial/permit_docker.bats#L73

Postscreen:
https://github.com/docker-mailserver/docker-mailserver/blob/bd96c1161e9f9b32bb0b3aaec4b930980a56945f/test/tests/parallel/set1/spam_virus/postscreen.bats#L40-L61

### Postgrey

Log stdout to service specific file (_replace [these two lines](https://github.com/docker-mailserver/docker-mailserver/blob/bd96c1161e9f9b32bb0b3aaec4b930980a56945f/target/supervisor/conf.d/supervisor-app.conf#L86-L87), this is the only service that lacks that_):

```
stdout_logfile=/var/log/supervisor/%(program_name)s.log
stderr_logfile=/var/log/supervisor/%(program_name)s.log
```

Replace these two lines with `"${LOGFILE_POSTGREY}"`

https://github.com/docker-mailserver/docker-mailserver/blob/bd96c1161e9f9b32bb0b3aaec4b930980a56945f/test/tests/parallel/set1/spam_virus/postgrey_enabled.bats#L64

https://github.com/docker-mailserver/docker-mailserver/blob/bd96c1161e9f9b32bb0b3aaec4b930980a56945f/test/tests/parallel/set1/spam_virus/postgrey_enabled.bats#L126

**Comment by georglauterbach** [comment]
I like this and will assist with integration.

**Comment by polarathene** [comment]
> I like this and will assist with integration.

Awesome, no rush as I'm not sure if it's the right time to bring it in.
- I need to find my Dovecot  Quota tests PR (_on a backup somewhere since my PC died earlier this year_), and probably finish wrangling the test suite conversion 😅 
- I may be able to dig up earlier Vector config from that experiment I did earlier this year, but I'm even less confident in where that lives (_if it's still on a disk_). I don't think it was too complicated, so you can probably get a similar setup after going through Vectors docs for a bit.
- We may want the Syslog sink feature to get merged/upstreamed first.
- Optimizing binary size shouldn't be too much trouble, if we take that route we could just use a separate Dockerfile / stage for now and `COPY --link` the binary over. Alternatively a separate repo could have CI automate Github Releases or publish a container of a lightweight build that we source it from. Or we hand wave away the size concern (I think it was 50MB+?)

**Comment by polarathene** [comment]
Review feedback that heavily documents logging config in DMS (_related to Debian 11 => Debian 12 `/etc/rsyslog.conf` change_):

- https://github.com/docker-mailserver/docker-mailserver/pull/3854#discussion_r1470525809
- https://github.com/docker-mailserver/docker-mailserver/pull/3854#discussion_r1470599989

An earlier comment above [has referenced this resource](https://github.com/docker-mailserver/docker-mailserver/pull/3134#issuecomment-1449598045), but it's worth noting here for me. Notably the implicit Amavis syslog debian config.

---

Aug 2024 update:

Related to the v14 release with `rsyslog` changes (_as per our `CHANGELOG.md` for v14_), as the log format changed I documented some useful information here (_notably regarding `logger`_): https://github.com/docker-mailserver/docker-mailserver/pull/4035#pullrequestreview-2079576115

While that notes issues with `logger` with the two syslog RFC format options, if neither of those are provided the log lines generated seemed compatible with what other services were generating:

```console
$ logger --priority mail.error --id=1337 --tag dms-changedetector "This is an example error message!"

# It would be processed via rsyslog to a log line like this (but with the timestamp prepended):
mail dms-changedetector[1337]: This is an example error message!
```

That ends up in the filtered `/var/log/mail.log` (symlink) / `/var/log/mail/mail.log` (_created by `rsyslogd`, with fallback via  `touch` in `start-mailserver.sh`_) and in this case also `/var/log/syslog`. I noticed that when the facility was changed from `mail` to another like `cron` or `uucp` that it was still being adjusted to `mail` instead of keeping the intended facility.. Potentially a misconfiguration with rsyslog?

For reference we don't seem to do much with rsyslog config, but it's manipulated a bit in the `Dockerfile`:

https://github.com/docker-mailserver/docker-mailserver/blob/2f8ad142eccfe3d6291373629c5b1dbf89b71f52/Dockerfile#L185-L188

https://github.com/docker-mailserver/docker-mailserver/blob/2f8ad142eccfe3d6291373629c5b1dbf89b71f52/Dockerfile#L242-L243

https://github.com/docker-mailserver/docker-mailserver/blob/2f8ad142eccfe3d6291373629c5b1dbf89b71f52/Dockerfile#L248-L249

https://github.com/docker-mailserver/docker-mailserver/blob/2f8ad142eccfe3d6291373629c5b1dbf89b71f52/Dockerfile#L262-L266

Log management and config as a whole should probably be reorganized once this feature is ready to be tackled.

---

Another insight about log concerns with the project, specifically around logrotate and `/var/log/mail` with `syslog:root` ownership: https://github.com/docker-mailserver/docker-mailserver/pull/4370#issuecomment-2661675385

---

For `logwatch` service I've documented an overview of our integration with that service [here](https://github.com/orgs/docker-mailserver/discussions/4539#discussioncomment-13996614), might be worth looking into relevance with this feature if replacing rsyslog or anything else that is writing logs it captures?
